src/controller/python/chip/discovery/__init__.py

```diff
@@ -0,0 +1,213 @@
+
+  def OnSuccess(self, peerId: PeerId, address: NodeAddress):
+    """Notify of a succesful address resolution."""
+    with self.operationCondition:
+      for item in self.activeDiscoveries:
+        if item.result.peerId != peerId:
+          continue
+      
+        item.result.addresses.add(address)
+        if item.firstResultTime == 0:
+          item.firstResultTime = time.time()
+      self.operationCondition.notify()
+      
+  
+  def ResolutionThread(self):
+    while True:
+      with self.operationCondition:
+        self.operationCondition.wait(self.ComputeNextEventTimeoutSeconds())
+
+        updatedDiscoveries = []
+
+        for item in self.activeDiscoveries:
+          if self.NeedsCallback(item):
+            try:
+              item.callback(item.result)
+            except:
+              logging.exception("Node discovery callback failed")
+          else:
+            updatedDiscoveries.append(item)
+        
+        self.activeDiscoveries = updatedDiscoveries
+
+  def NeedsCallback(self, item: PendingDiscovery):
+    """Find out if the callback needs to be called for the given item."""
+    now = time.time()
+    if item.expireTime <= now:
+      return True
+    
+    if (item.firstResultTime > 0) and (item.firstResultTime + _RESULT_WAIT_TIME_SEC <= now):
+      return True
+
+    return False
+
+  def ComputeNextEventTimeoutSeconds(self):
+    """Compute how much a thread needs to sleep based on the active discoveries list."""
+    sleepTimeSec = 10.0  # just in case
+
+    now = time.time()
+    for item in self.activeDiscoveries:
+      # figure out expiry time for an item
+      expireSleep = item.expireTime - now 
+      if expireSleep < sleepTimeSec:
+        sleepTimeSec = expireSleep
+
+      # Allow only a short time window for 'additional results' once we 
+      # have one
+      resultSleep = (item.firstResultTime + _RESULT_WAIT_TIME_SEC) - now
+      if resultSleep < sleepTimeSec:
+        sleepTimeSec = resultSleep
+      
+    
+    # in case our processing missed some expire times, set a very short
+    # sleep
+    if sleepTimeSec <= 0:
+      sleepTimeSec = 0.001
+
+    return sleepTimeSec
+
+
+# define firstResultTime
+
+
+# All pending discovery operations awayting callback results
+_gPendingDiscoveries = _PendingDiscoveries()
+
+
+@DiscoverSuccessCallback_t
+def _DiscoverSuccess(fabric: int, node: int, interface: int, ip: str,  port: int):
+  peerId = PeerId(fabric, node)
+  address = NodeAddress(interface, ip, port)
+
+  global _gPendingDiscoveries
+  _gPendingDiscoveries.OnSuccess(peerId, address)
+
+
+@DiscoverFailureCallback_t
+def _DiscoverFailure(fabric: int, node: int, errorCode: int):
+  # Many discovery errors currently do not include a useful node/fabric id 
+  # hence we just log and rely on discovery timeouts to return 'no data'
+  logging.error("Discovery failure, error %d", errorCode)
+
+
+def FindAddressAsync(fabricid, nodeid, callback, timeout_ms=1000):
```

Comment:
missing type annotations?

Reply:
Added, however I do not know offhand how to properly type-annotate callbacks (will do this once back in the office).